src/core/mavsdk_impl.cpp

```diff
@@ -239,11 +232,27 @@ bool MavsdkImpl::send_message(mavlink_message_t& message)
 {
     std::lock_guard<std::mutex> lock(_connections_mutex);
 
+    uint8_t successful_emissions = 0;
     for (auto it = _connections.begin(); it != _connections.end(); ++it) {
-        if (!(**it).send_message(message)) {
-            LogErr() << "send fail";
-            return false;
+        // Checks whether connection knows target system ID by extracting target system if set.
+        // https://github.com/PX4/PX4-Autopilot/blob/v1.11.3/src/modules/mavlink/mavlink_main.cpp#L472
+        const mavlink_msg_entry_t* meta = mavlink_get_msg_entry(message.msgid);
+
+        if (meta && meta->flags & MAV_MSG_ENTRY_FLAG_HAVE_TARGET_SYSTEM) {
+            int target_system_id = (_MAV_PAYLOAD(&message))[meta->target_system_ofs];
+            if (!(**it).has_system_id(target_system_id)) {
+                continue;
+            }
         }
```

Comment:
Just for my understanding: this is an optimization, right? It won't send a targeted message to a connection that has not been seen to have this sysid, but a broadcast will be sent. Correct?

Or is it required for some other reason? If it is just an optimization, I'm not sure if it is worth keeping, and actually for forwarding we may need to handle it differently to avoid loops, right?

Reply:
Yes, it is meant to be an optimization. A broadcast should be sent as I registered the "virtual" system ID 0 in the `Connection` constructor.
But I am not sure that the following line returns 0 when it is not a targeted message:
```
 int target_system_id = (_MAV_PAYLOAD(&message))[meta->target_system_ofs];

```

About its utility, let's consider an exemple, if we have an autopilot **(FC,1.1)** linked to a Companion Computer **(CC,1.195)** which is connected to a Ground Station **(GS,245.190)** and another device **(O,12.195)**, all using `mavsdk`:
- FC <--> CC = Serial
- CC <--> GS = Radio
- CC <--> O = Radio

Let's say the **GS** sends messages at high rate to the **FC** through the **CC**.
If we do not filter messages, the **CC** will forward messages targeted to the **FC** with system ID 1 to both the **FC** and **O**, even though **O** will ignore it. Indeed, the `Connection` linking **CC** to **O** will never receive any message from the sys ID 1, so it will not register it. Thus, the (radio) bandwidth will be loaded unnecessarily.

By the way, you are right for the loop problem.
When a `mavsdk` instance forward a message, it ignores the `Connection` from which it received the message.
However, if the network contains a loop, the message will be forwarded forever.
In the example above, if we add the following link:
- FC <--> GS = Radio (or whatever)
Then we create a loop: GS -> CC -> FC -> GS -> CC -> FC -> ...

Reply:
Got it, good to know that it is an optimization. Other than that, I think that it doesn't hurt, so if it works, it can stay :+1:.

Reply:
All right ! 😄 
Maybe a comment should be added to prevent user from creating such loop when activating forwarding option ...